### PR TITLE
feat(FileList): [TEA-2545] Implement internal file list component

### PR DESCRIPTION
**Pull request by besirovic**
[TEA-2545]

[DESIGN](https://share.goabstract.com/2556a122-6443-4546-96db-3d40c37b2a0d?mode=build&sha=0aeb05e01da837c228249c74cc79ec21280f8b6d)

### Description

Implement internal `FileList` component needed for `FileInput` component. Component will be added to `FileInput` as part of  [TEA-2546](https://toptal-core.atlassian.net/browse/TEA-2546).

> NOTE
>
> `FileList` example inside `FileInput` section is added just for testing purpose and will be removed as part of  [TEA-2546](https://toptal-core.atlassian.net/browse/TEA-2546)

### How to test

- In storybook, open `FileInput` component and find `FileList` section on the bottom
- Check if list of files is implemented according design

### Screenshots

Screenshot 2021-05-12 at 02 14 06

### Review

- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/api-principles.md)
- [x] Annotate all `props` in component with documentation
- [x] Create `examples` for component
- [ ] Ensure that deployed demo has expected results and good examples
- [x] Ensure that tests pass by running `yarn test`
- [ ] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

[TEA-2545]: https://toptal-core.atlassian.net/browse/TEA-2545

### CODE REVIEW

File: packages/picasso/src/FileInput/story/FileList.example.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import React from 'react'
+import { FileList } from '@toptal/picasso'
+
+const Example = () => {
```
Comment (besirovic) [inline packages/picasso/src/FileInput/story/FileList.example.tsx:4]: As mentioned in description of PR, this example is added just for testing purpose and will be removed as part of  [TEA-2546](https://toptal-core.atlassian.net/browse/TEA-2546).

File: packages/picasso/src/FileInput/story/index.jsx [hunk 2]
```diff
@@ -25,3 +25,7 @@ page
     description:
       'Showcase how to integrate FileInput with uploading logic. Randomly returns errors and allows only upload of image files'
   })
+  .addExample('FileInput/story/FileList.example.tsx', {
+    title: 'FileList',
+    description: 'List of files'
+  })
```
Comment (besirovic) [inline packages/picasso/src/FileInput/story/index.jsx:31]: As mentioned in description of PR, this example is added just for testing purpose and will be removed as part of  [TEA-2546](https://toptal-core.atlassian.net/browse/TEA-2546).

**Comment by besirovic** [comment]
@toptal-bot run all

File: packages/picasso/src/FileList/FileList.tsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+export interface Props extends BaseProps, HTMLAttributes<HTMLDivElement> {
+  files: File[]
+  onItemRemove?: (fileName: string) => void
+}
+
+const useStyles = makeStyles<Theme>(styles, {
+  name: 'FileList'
+})
+
+const FileListItem = ({ file, onRemove }: FileListItemProps) => {
```
Comment (teimurjan) [inline packages/picasso/src/FileList/FileList.tsx:35]: We're trying to keep one component per file. It'd be awesome if you move it to a separate module with its own styles and tests.

File: packages/picasso/src/FileList/FileList.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  const { uploading, progress = 0, error, name } = file
+
+  const classes = useStyles()
+
+  const renderUploading = (
```
Comment (teimurjan) [inline packages/picasso/src/FileList/FileList.tsx:40]: It's not a function, so it shouldn't contain a verb in its name.

```suggestion
  const renderUploading = () => (
```

or 

```suggestion
  const uploadingNode = (
```

File: packages/picasso/src/FileList/FileList.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    <>
+      <Typography variant='body' color='black' size='medium'>
+        Uploading...
+      </Typography>
+      <ProgressBar value={progress} data-testid='file-list-item-progressbar' />
+    </>
+  )
+
+  const renderFile = (
```
Comment (teimurjan) [inline packages/picasso/src/FileList/FileList.tsx:49]: Fix the naming here, too, please.

File: packages/picasso/src/FileList/FileList.tsx [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            {name}
+          </Typography>
+        </Container>
+        <Typography variant='body' size='small' color='red'>
+          {error}
+        </Typography>
+      </Container>
+      {onRemove && (
+        <Tooltip compact placement='top' variant='dark' content='Remove file'>
+          <Button.Circular variant='flat' onClick={() => onRemove(name)}>
```
Comment (teimurjan) [inline packages/picasso/src/FileList/FileList.tsx:72]: Let's extract the inline function.

```suggestion
          <Button.Circular variant='flat' onClick={handleRemove}>
```

File: packages/picasso/src/FileList/test.tsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+describe('FileList', () => {
+  const file = {
+    uploading: false,
+    progress: 0,
+    error: undefined,
+    name: 'user-profile-picture.png'
+  }
+
+  it('renders', () => {
+    const { container } = render(<FileList files={[file]} />)
```
Comment (teimurjan) [inline packages/picasso/src/FileList/test.tsx:15]: It'd be great if you create a render function with the default arguments like this:

https://github.com/toptal/picasso/blob/0d61c462669455c207854f380ebe8963688841a4/packages/picasso/src/Accordion/test.tsx#L22

We're following this pattern in the codebase.

File: packages/picasso/src/FileList/test.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+      const { getByRole } = render(
+        <FileList files={[file]} onItemRemove={handleIItemRemove} />
+      )
+
+      fireEvent.click(getByRole('button'))
+
+      expect(handleIItemRemove).toHaveBeenCalled()
+    })
+  })
```
Comment (teimurjan) [inline packages/picasso/src/FileList/test.tsx:32]:
```suggestion
it('fires callback on remove', () => {
  const handleIItemRemove = jest.fn()

  const { getByRole } = render(
    <FileList files={[file]} onItemRemove={handleIItemRemove} />
  )

  fireEvent.click(getByRole('button'))

  expect(handleIItemRemove).toHaveBeenCalled()
})
```

**Review by teimurjan — CHANGES_REQUESTED** [review]

File: packages/picasso/src/FileListItem/FileListItem.tsx [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+const FileListItem = ({ file, onRemove }: Props) => {
+  const { uploading, progress = 0, error, name } = file
+
+  const classes = useStyles()
+
+  const handleRemove = () => {
+    if (onRemove !== undefined) {
+      onRemove(name)
+    }
```
Comment (teimurjan) [inline packages/picasso/src/FileListItem/FileListItem.tsx:31]:
```suggestion
    onRemove?.(name)
```

**Review by teimurjan — APPROVED** [review]

**Review by eduardoportilho — APPROVED** [review]

**Pull request merged**